docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+Writing Plugins
+========================
+
+- [Introduction](#introduction)
+- [Writing a plugin](#writing-a-plugin)
+  - [Setting up your plugin](#setting-up-your-plugin)
+  - [Providing generator functions](#providing-generator-functions)
+     - [Overriding transpilation](#overriding-transpilation) 
+  - [Generating a file](#generating-a-file)
+  - [Walking through the schema](#walking-through-the-schema)
+  - [Printing to a generated file](#printing-to-a-generated-file)
+  - [Importing](#importing)
+     - [Importing from an NPM package](#importing-from-an-npm-package) 
+     - [Importing from `protoc-gen-es` generated code](#importing-from-protoc-gen-es-generated-code) 
+     - [Importing from the `@bufbuild/protobuf` runtime](#importing-from-the-bufbuildprotobuf-runtime)
+     - [Type-only imports](#type-only-imports)
+     - [Why use `f.import()`?](#why-use-fimport)
+  - [Exporting](#exporting)
+  - [Parsing plugin options](#parsing-plugin-options)
+  - [Reading custom options](#reading-custom-options)
+     - [Scalar options](#scalar-options) 
+     - [Message options](#message-options) 
+     - [Enum options](#enum-options) 
+- [Testing](#testing)
+- [Examples](#examples)
+
+## Introduction
+
+Code generator plugins are a unique feature of protocol buffer compilers like protoc and the [buf CLI](https://docs.buf.build/introduction#the-buf-cli).  With a plugin, you can generate files based on Protobuf schemas as the input.  Outputs such as RPC clients and server stubs, mappings from protobuf to SQL, validation code, and pretty much anything else you can think of can all be produced.
+
+The contract between the protobuf compiler and a code generator plugin is defined in [plugin.proto](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/compiler/plugin.proto). Plugins are simple executables (typically on your `$PATH`) that are named `protoc-gen-x`, where `x` is the name of the language or feature that the plugin provides. The protobuf compiler parses the protobuf files, and invokes the plugin, sending a `CodeGeneratorRequest` on standard in, and expecting a `CodeGeneratorResponse` on standard out. The request contains a set of descriptors (see [descriptor.proto](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/descriptor.proto)) - an abstract version of the parsed protobuf files. The response contains a list of files, each having a name and text content.
+
+For more information on how plugins work, check out [our documentation](https://docs.buf.build/reference/images). 
+
+## Writing a plugin
+
+The following will describe the steps and things to know when writing your own code generator plugin.  The main step in the process is passing a plugin initialization object to the `createEcmaScriptPlugin` function exported by the plugin framework.  This plugin initalization object will contain various properties pertaining to different aspects of your plugin.
```

Comment:
Would be helpful to call out how to find the plugin framework with a link to the package on npmjs.com.